### PR TITLE
8314550: [macosx-aarch64] serviceability/sa/TestJmapCore.java fails with "sun.jvm.hotspot.debugger.UnmappedAddressException: 801000800"

### PR DESCRIPTION
**Pull request by plummercj**
On some macosx-aarch64 systems, not all mapped pages are dumped to the core file. This first turned up with [JDK-8293563](https://bugs.openjdk.org/browse/JDK-8293563) where large parts of the ZGC heap would not be in the core file, leading to various SA address errors. For JDK-8293563 the issue was addressed by having the core file test always use `-XX:+AlwaysPreTouch` on macosx-aarch64. This seemed to force the ZGC pages to always end up in the core file.

A similar issue has been noticed with mapped in pages of the CDS archive. We are seeing cases where SA references to addresses that are clearly in the CDS archive (based on info in the hs_err file) are failing to be read from the core file by SA. This problem has turned up a number of times during CI testing, but I have yet to be able to reproduce it myself. This PR is an attempt to address this testing issue by having the CDS archive also pretouch all mapped in pages when `-XX:+AlwaysPreTouch` is used.

Tested with tier1 and tier3 and also ran the test about 5,000 times with and without the fix. It never reproduced for either. Hopefully the problem is gone with this fix, but it may take a few months of CI testing before we can be confident it is fixed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314550](https://bugs.openjdk.org/browse/JDK-8314550): [macosx-aarch64] serviceability/sa/TestJmapCore.java fails with "sun.jvm.hotspot.debugger.UnmappedAddressException: 801000800" (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [72b6fbcd](https://git.openjdk.org/jdk/pull/15423/files/72b6fbcd45fb8ae2ac1241cb0979b58595f39e50)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [72b6fbcd](https://git.openjdk.org/jdk/pull/15423/files/72b6fbcd45fb8ae2ac1241cb0979b58595f39e50)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [72b6fbcd](https://git.openjdk.org/jdk/pull/15423/files/72b6fbcd45fb8ae2ac1241cb0979b58595f39e50)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15423/head:pull/15423` \
`$ git checkout pull/15423`

Update a local copy of the PR: \
`$ git checkout pull/15423` \
`$ git pull https://git.openjdk.org/jdk.git pull/15423/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15423`

View PR using the GUI difftool: \
`$ git pr show -t 15423`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15423.diff">https://git.openjdk.org/jdk/pull/15423.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15423#issuecomment-1692549641)

### CODE REVIEW

File: src/hotspot/share/cds/filemap.cpp [hunk 1]
```diff
@@ -1683,6 +1683,22 @@ void FileMapInfo::close() {
...
 
+/*
+ * Same as os::map_memory() but also pretouches if AlwaysPreTouch is enabled.
+ */
+char* map_memory(int fd, const char* file_name, size_t file_offset,
+                 char *addr, size_t bytes, bool read_only,
+                 bool allow_exec, MEMFLAGS flags = mtNone) {
+  if (AlwaysPreTouch) {
+    read_only = false;
+  }
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/filemap.cpp:1694]: Nit: you can handle this more succinctly by simply doing:
```
 char* mem = os::map_memory(fd, file_name, file_offset, addr, bytes,
                            AlwaysPreTouch ? false : read_only, allow_exec, flags);
```

**Review by dholmes-ora — COMMENTED** [review]
As `AlwaysPreTouch` is false by default, and we don't explicitly turn it on for CI testing, what tests will actually be affected by this change?  And for applications that already enable `AlwaysPreTouch` (would they?) what impact might this have e.g. on startup?

Thanks

**Comment by plummercj** [comment]
> As `AlwaysPreTouch` is false by default, and we don't explicitly turn it on for CI testing, what tests will actually be affected by this change? And for applications that already enable `AlwaysPreTouch` (would they?) what impact might this have e.g. on startup?

In the CR I mentioned the following was added as part of the fix done in [JDK-8293563](https://bugs.openjdk.org/browse/JDK-8293563) (which solves this same issue for the ZGC heap):

     public static String getAlwaysPretouchArg(boolean withCore) {
        // macosx-aarch64 has an issue where sometimes the java heap will not be dumped to the
        // core file. Using -XX:+AlwaysPreTouch fixes the problem.
        if (withCore && Platform.isOSX() && Platform.isAArch64()) {
            return "-XX:+AlwaysPreTouch";
        } else {
            return "-XX:-AlwaysPreTouch";
        }
    } 

So this bit of code will make sure that all the SA core file tests use AlwaysPreTouch on macosx-aarch64.

As for performance impact. I'm not certain and I'm also not to certain of the use case for AlwaysPretouch. As far as I can tell in was introduced by [JDK-6806090](https://bugs.openjdk.org/browse/JDK-6806090) and seems it was actually added to help with performance, although I'm not sure how.

The CDS archive tends to be small compared with the java heap, so I would think the relative impact would be small. Also unclear is how much of the archive is likely to get mapped in on startup anyway. Perhaps @iklam can add some insight.

**Review by iklam — APPROVED** [review]
Looks good to me.

The size of the CDS archive is usually much smaller than the committed heap size. 

The [GC tuning docs says](https://docs.oracle.com/en/java/javase/11/gctuning/garbage-first-garbage-collector-tuning.html#GUID-0770AB01-E334-4E23-B307-FD2114B16E0E):

> You can use -XX:+AlwaysPreTouch to move the operating system work to back virtual memory with physical memory to VM startup time. Both of these measures can be particularly desirable in order to make pause-times more consistent.

So committing all the CDS regions to physical memory at VM start-up would be what the user expects when `-XX:+AlwaysPreTouch` is specified.

File: src/hotspot/share/cds/filemap.cpp [hunk 2]
```diff
@@ -1683,6 +1683,22 @@ void FileMapInfo::close() {
...
+  char* mem = os::map_memory(fd, file_name, file_offset, addr, bytes,
+                             read_only, allow_exec, flags);
+  if (mem != nullptr && AlwaysPreTouch) {
+    os::pretouch_memory(mem, mem + bytes, os::vm_page_size());
```
Comment (tstuefe) [inline src/hotspot/share/cds/filemap.cpp:1698]: Nit: you can omit the page size since it defaults to os::vm_page_size.

**Review by tstuefe — APPROVED** [review]
LGTM

**Review by dholmes-ora — APPROVED** [review]
Thanks for clarifying @plummercj . LGTM.

Thanks

**Comment by plummercj** [comment]
Thanks for the reviews Ioi, Thomas, and David!

/integrate

**Pull request closed without merging**